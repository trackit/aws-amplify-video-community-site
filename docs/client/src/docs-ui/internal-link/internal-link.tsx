import {Component, h, Prop, Host, State, Watch, Build} from "@stencil/core";
import {filtersByRoute} from "../../api";
import {SelectedFilters} from "../page/page.types";
import {pageContext} from "../page/page.context";
import {getPage} from "../../cache.worker";
import {parseURL, serializeURL} from "../../utils/url/url.worker";

@Component({tag: "docs-internal-link", shadow: false})
export class DocsInternalLink {
  /*** the global selected filter state */
  @Prop() readonly selectedFilters?: SelectedFilters;
  /*** the route to render out */
  @Prop() readonly href: string;
  /*** query string parameters to attach to the link */
  @Prop() readonly QSPs: string = "";
  /*** class name to attach to link when active */
  @Prop() readonly activeClass?: string;
  /*** class name to attach a subpage is active */
  @Prop() readonly childActiveClass?: string;
  /*** override `isChildActive` to true */
  @Prop() readonly additionalActiveChildRoots?: string[];

  selectedFilter?: string;

  @State() url?: string;
  @State() isActive?: boolean;
  @State() isChildActive?: boolean;

  @Watch("selectedFilters")
  async componentWillLoad() {
    let selectedFilter: string | undefined;
    if (Build.isBrowser && !this.href) return;
    const {path, hash, params} = await parseURL(this.href);

    if (Object.keys(params).length === 0) {
      const filters = filtersByRoute.get(path);
      if (filters) {
        const [[filterKey, filterValues]] = Object.entries(filters);
        selectedFilter = this.selectedFilters?.[filterKey] as
          | string
          | undefined;
        if (selectedFilter) {
          if (
            Array.isArray(filterValues) &&
            filterValues.includes(selectedFilter)
          ) {
            params[filterKey] = selectedFilter;
          }
        }
      }
    }

    const url = await serializeURL({path, hash, params});
    const isActive = location.pathname === url;
    const currentPathWithoutQS = location.pathname.split("/q/")?.[0];
    const hrefWithoutQS = url.split("/q/")?.[0];
    const isChildActive =
      this.additionalActiveChildRoots?.some((root) =>
        location.pathname.startsWith(root),
      ) ||
      !!(
        hrefWithoutQS &&
        currentPathWithoutQS?.startsWith(hrefWithoutQS) &&
        !currentPathWithoutQS?.startsWith(`${hrefWithoutQS}-`)
      );

    Object.assign(this, {url, isActive, isChildActive});
  }

  async componentDidRender() {
    if (this.url) {
      getPage((await parseURL(this.url)).path);
    }
  }

  render() {
    return (
      <Host>
        <stencil-route-link
          url={this.url + this.QSPs}
          class={{
            ...(this.activeClass ? {[this.activeClass]: !!this.isActive} : {}),
            ...(this.childActiveClass
              ? {[this.childActiveClass]: !!this.isChildActive}
              : {}),
          }}
        >
          <slot />
        </stencil-route-link>
      </Host>
    );
  }
}

pageContext.injectProps(DocsInternalLink, ["selectedFilters"]);
