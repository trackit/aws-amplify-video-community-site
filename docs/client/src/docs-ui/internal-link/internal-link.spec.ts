import {DocsInternalLink} from "./internal-link";
import {newSpecPage} from "@stencil/core/testing";

describe("docs-internal-link", () => {
  describe("Component logic", () => {
    let internalLink: DocsInternalLink;
    beforeEach(() => (internalLink = new DocsInternalLink()));

    it("should init `href` as `undefined`", () =>
      expect(internalLink.href).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsInternalLink],
            html: `<docs-internal-link />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
