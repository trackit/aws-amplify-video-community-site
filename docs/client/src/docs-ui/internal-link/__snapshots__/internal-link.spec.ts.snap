// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs-internal-link Render logic should render 1`] = `
<docs-internal-link>
  <!---->
  <stencil-route-link url="/undefined"></stencil-route-link>
</docs-internal-link>
`;
